docs/content/concepts/partitions-schedules-sensors/sensors.mdx

```diff
@@ -205,6 +191,58 @@ def every_fifth_asset_c_sensor(context):
         return SkipReason(f"asset_c only materialized {len(asset_events)} times.")
 ```
 
+#### Triggering a partitioned run after upstream partitions are materialized
+
+The <PyObject object="MultiAssetSensorEvaluationContext" method="latest_materialization_records_by_partition"/> method fetches the latest materialization records by partition for a given asset. Using this method, you can trigger downstream partitioned runs when the corresponding partitions of upstream assets are materialized.
+
+The following example monitors two upstream daily-partitioned assets, kicking off materializations of the corresponding partition in the downstream daily-partitioned asset. Whenever a partition is replaced in `upstream_daily_1` and the same partition has previously been materialized in `upstream_daily_2` (and vice versa), the sensor yields a run request for the downstream asset.
+
+```python file=/concepts/partitions_schedules_sensors/sensors/sensors.py startafter=start_daily_partitioned_asset_with_two_upstreams endbefore=end_daily_partitioned_asset_with_two_upstreams
+@multi_asset_sensor(
+    asset_keys=[AssetKey("upstream_daily_1"), AssetKey("upstream_daily_2")],
+    job=downstream_daily_job,
+)
+def trigger_daily_asset_if_all_upstream_partitions_materialized(context):
```

Comment:
This is a pretty complex implementation for such a common use case. I wonder if there's additional utility methods we could supply to make it simpler?

Reply:
Hm... taking into account your other feedback, here's an updated version of this sensor:
```
@multi_asset_sensor(
    asset_keys=[AssetKey("upstream_daily_1"), AssetKey("upstream_daily_2")],
    job=downstream_daily_job,
)
def trigger_daily_asset_if_all_upstream_partitions_materialized(context):
    run_requests_by_partition = {}

    for asset_key in context.asset_keys:
        for (
            partition,
            materialization,
        ) in context.latest_materialization_records_by_partition(asset_key).items():
            if (
                all(
                    [
                        context.all_partitions_materialized(other_key, [partition])
                        for other_key in context.asset_keys
                        if other_key != asset_key
                    ]
                )
                and partition not in run_requests_by_partition
            ):
                run_requests_by_partition[
                    partition
                ] = downstream_daily_job.run_request_for_partition(partition)
            context.advance_cursor({asset_key: materialization})

    return list(run_requests_by_partition.values())
```
One small optimization I can think of is to allow `all_partitions_materialized` to accept a list of asset keys, or use all monitored asset keys by default. But aside from that, it's hard to think of more because we are constrained to advancing the cursor for each monitored asset key sequentially

Reply:
What if we had a method that returned a dictionary that mapped each partition key to a list or set of asset keys that had materializations with that partition? (We would probably want to exclude partitions with no materializations or something).  Then the user could just iterate through that list?

Reply:
Yeah, we could do that. If we did that we'd have to advance all of the cursors in that call (unless if we plan on passing the materialization objects to the user within the call).

Reply:
> (unless if we plan on passing the materialization objects to the user within the call).

Is there a downside to doing that? I think in general it's better to separate the methods that return information from the methods that advance the cursor.

Reply:
I think the only downside is that the returned object gets a little bit more complicated, it would be `Mapping[str, Mapping[AssetKey, List[EventLogEntry]]]`

Reply:
Agree that we want to avoid advancing the cursor in this call though

Reply:
Could it just be `Mapping[str, Tuple[AssetKey, EventLogEntry]`? I.e. only return the latest entry for that asset-partition combo?

Reply:
Yeah, we can do that.

One nuance with doing something like this is that the materializations will be returned out of order (from the order they occurred). So if we provided this method, users would have to use `context.advance_all_cursors` instead of `context.advance_cursor`.

For example, if asset A had partition A materialized and then partition B, and asset B had partition B materialized and then partition A, this method would return:
```
{
  partition A: [(Asset A, partition a materialization), (Asset B, partition B materialization)],
  partition B: [(asset A, partition b materialization), (Asset B, partition B materialization)]
}
And so calling `advance_cursor` (instead of advance_all_cursors) for partition B after partition A would backtrack the cursor for partition B